File: csharp/src/exercises/code/BinaryTreeLevelOrderTraversal.cs
namespace exercises.code
{
    // Definição do nó da árvore binária
    public class TreeNode
    {
        public int val;         
        public TreeNode left;   
        public TreeNode right;    

        public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
        {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public class Solution
    {
        // Realiza a travessia em nível (level-order) da árvore binária
        public IList<IList<int>> LevelOrder(TreeNode root)
        {
            // Lista de listas para armazenar os valores por nível
            var res = new List<IList<int>>();

            // Se a raiz for nula, retorna lista vazia
            if (root == null) return res;

            // Fila para armazenar os nós que serão visitados
            var queue = new Queue<TreeNode>();
            queue.Enqueue(root); // Começa pela raiz

            // Enquanto houver nós na fila
            while (queue.Count > 0)
            {
                int levelSize = queue.Count;     // Número de nós no nível atual
                var level = new List<int>();     // Lista para armazenar os valores deste nível

                // Processa todos os nós do nível atual
                for (int i = 0; i < levelSize; i++)
                {
                    var node = queue.Dequeue();  // Remove o nó da fila
                    level.Add(node.val);         // Adiciona o valor do nó à lista do nível

                    // Se o nó possui filho à esquerda, adiciona na fila
                    if (node.left != null)
                        queue.Enqueue(node.left);

                    // Se o nó possui filho à direita, adiciona na fila
                    if (node.right != null)
                        queue.Enqueue(node.right);
                }

                // Após processar o nível, adiciona a lista à resposta final
                res.Add(level);
            }

            // Retorna a lista com os valores por nível
            return res;
        }
    }
}

File: go/exercises/code/BinaryTreeLevelOrderTraversal.go
package code

// TreeNode representa um nó da árvore binária
/*
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
*/

// levelOrder realiza a travessia por nível (level-order / BFS) de uma árvore binária.
// Retorna uma matriz de inteiros, onde cada linha representa os valores de um nível da árvore.
func LevelOrder(root *TreeNode) [][]int {
	var res [][]int // Resultado final: uma matriz de níveis

	// Se a árvore estiver vazia, retorna uma matriz vazia
	if root == nil {
		return res
	}

	// Fila para armazenar os nós a serem processados
	queue := []*TreeNode{root}

	// Enquanto houver nós na fila
	for len(queue) > 0 {
		levelSize := len(queue) // Número de nós no nível atual
		var level []int         // Lista para armazenar os valores deste nível

		// Processa todos os nós do nível atual
		for i := 0; i < levelSize; i++ {
			node := queue[0]       // Pega o primeiro nó da fila
			queue = queue[1:]      // Remove o nó da fila

			level = append(level, node.Val) // Adiciona o valor do nó atual ao nível

			// Adiciona os filhos do nó atual à fila (para o próximo nível)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		// Adiciona o nível completo à resposta final
		res = append(res, level)
	}

	// Retorna todos os níveis da árvore
	return res
}
